onnx/shape_inference/implementation.cc

```diff
@@ -684,6 +688,40 @@ void InferShapes(
   }
 }
 
+std::unordered_map<std::string, TensorShapeProto> InferShapesAndDataPropagation(
+    ModelProto& m,
+    const std::unordered_map<std::string, TensorShapeProto>& generated_shape_data_by_name,
+    const ISchemaRegistry* schema_registry,
+    const ShapeInferenceOptions& options) {
+  std::unordered_map<std::string, int> opset_imports;
```

Comment:
I wonder why some of the code below doesn't happen in the constructor of ShapeInferenceImpleBase? May be with a separate, different, constructor from the existing one?

Reply:
I think the code below are only common in 2 places: one is in ` InferShapes` and another one is `InferShapesAndDataPropagation`, because only these 2 take ModelProto as input. Although we cannot put them into ShapeInferenceImpleBase (it's for GraphProto), perhaps at least we can make them as a common function to reduce duplicate code.

Still, it seems that `TraverseGraphsToAddExistingSymbol` can be included in the ShapeInferenceImpleBase constructor. I will give it a try.

Reply:
I introduced two functions, `GetOpsetImportsFromProto` and `GetModelLocalFunctionsMapFromModel`, to reduce duplicate code between `InferShapes` and `InferShapesAndDataPropagation`. In addition, I moved TraverseGraphsToAddExistingSymbol into process(GraphProto& graph) since it's required step to process a graph.

BTW, I might find a potential bug that it does not consider symbols in a FunctionProto. I created an issue to track that: https://github.com/onnx/onnx/issues/4191.